File: examples/csharp/Webhook.cs
// Dona webhook signature verification (Standard Webhooks) — base class library only.
//
// Headers on every delivery: webhook-id · webhook-timestamp · webhook-signature.
//   signature = space-delimited "v1,<base64 HMAC-SHA256(base64decode(secret after whsec_), id.ts.rawBody)>"
//   (two entries for 24 h after a secret rotation — accept if ANY matches)
//
// ⚠ Verify the RAW request bytes; re-serialised JSON never matches. Then dedupe on webhook-id for 24 h.
//
//   app.MapPost("/dona/webhooks", async (HttpRequest req) => {
//       using var ms = new MemoryStream(); await req.Body.CopyToAsync(ms);
//       try { DonaWebhook.Verify(secret, req.Headers["webhook-id"]!, req.Headers["webhook-timestamp"]!,
//                                req.Headers["webhook-signature"]!, ms.ToArray()); }
//       catch (DonaWebhookVerificationException) { return Results.Unauthorized(); }
//       return Results.NoContent();   // acknowledge fast; process asynchronously
//   });
using System.Security.Cryptography;
using System.Text;

namespace DonaExamples;

public sealed class DonaWebhookVerificationException(string message) : Exception(message);

public static class DonaWebhook
{
    public static readonly TimeSpan Tolerance = TimeSpan.FromSeconds(300);

    /// Throws DonaWebhookVerificationException unless the delivery is authentic and fresh.
    public static void Verify(string secret, string id, string timestamp, string signatures, byte[] rawBody, DateTimeOffset? now = null)
    {
        if (string.IsNullOrEmpty(id) || string.IsNullOrEmpty(timestamp) || string.IsNullOrEmpty(signatures))
            throw new DonaWebhookVerificationException("missing webhook-* headers");
        if (!long.TryParse(timestamp, out var ts))
            throw new DonaWebhookVerificationException("bad webhook-timestamp");
        var skew = (now ?? DateTimeOffset.UtcNow).ToUnixTimeSeconds() - ts;
        if (Math.Abs(skew) > Tolerance.TotalSeconds)
            throw new DonaWebhookVerificationException("timestamp outside tolerance");

        byte[] key;
        try { key = Convert.FromBase64String(secret.StartsWith("whsec_", StringComparison.Ordinal) ? secret[6..] : secret); }
        catch (FormatException) { throw new DonaWebhookVerificationException("secret is not base64 after whsec_"); }

        byte[] message = [.. Encoding.UTF8.GetBytes($"{id}.{timestamp}."), .. rawBody];
        var expected = HMACSHA256.HashData(key, message);
        foreach (var entry in signatures.Split(' '))
        {
            var comma = entry.IndexOf(',');
            if (comma < 0 || entry[..comma] != "v1") continue;
            byte[] got;
            try { got = Convert.FromBase64String(entry[(comma + 1)..]); }
            catch (FormatException) { continue; }
            if (CryptographicOperations.FixedTimeEquals(got, expected)) return;
        }
        throw new DonaWebhookVerificationException("no matching signature");
    }
}

File: examples/php/webhook.php
<?php
// Dona webhook signature verification (Standard Webhooks) — no dependencies.
//
// Headers on every delivery: webhook-id · webhook-timestamp · webhook-signature.
//   signature = space-delimited "v1,<base64 HMAC-SHA256(base64decode(secret after whsec_), id.ts.rawBody)>"
//   (two entries for 24 h after a secret rotation — accept if ANY matches)
//
// ⚠ Verify the RAW request bytes: file_get_contents('php://input'), never json_encode(json_decode(...)).
//   Then dedupe on webhook-id (the event id) for 24 h.
//
//   $raw = file_get_contents('php://input');
//   try {
//       dona_verify_webhook(getenv('DONA_WEBHOOK_SECRET'), [
//           'webhook-id' => $_SERVER['HTTP_WEBHOOK_ID'] ?? '',
//           'webhook-timestamp' => $_SERVER['HTTP_WEBHOOK_TIMESTAMP'] ?? '',
//           'webhook-signature' => $_SERVER['HTTP_WEBHOOK_SIGNATURE'] ?? '',
//       ], $raw);
//   } catch (DonaWebhookVerificationError) { http_response_code(401); exit; }
//   http_response_code(204); // acknowledge fast; process asynchronously
declare(strict_types=1);

const DONA_WEBHOOK_TOLERANCE_SECONDS = 300;

final class DonaWebhookVerificationError extends RuntimeException {}

/**
 * @param array<string,string> $headers webhook-id, webhook-timestamp, webhook-signature (lower-case keys)
 * @throws DonaWebhookVerificationError unless the delivery is authentic and fresh
 */
function dona_verify_webhook(string $secret, array $headers, string $rawBody, ?int $now = null): void
{
    $id = $headers['webhook-id'] ?? '';
    $ts = $headers['webhook-timestamp'] ?? '';
    $signatures = $headers['webhook-signature'] ?? '';
    if ($id === '' || $ts === '' || $signatures === '') {
        throw new DonaWebhookVerificationError('missing webhook-* headers');
    }
    if (!ctype_digit($ts)) {
        throw new DonaWebhookVerificationError('bad webhook-timestamp');
    }
    $now ??= time();
    if (abs($now - (int) $ts) > DONA_WEBHOOK_TOLERANCE_SECONDS) {
        throw new DonaWebhookVerificationError('timestamp outside tolerance');
    }
    $key = base64_decode(str_starts_with($secret, 'whsec_') ? substr($secret, 6) : $secret, true);
    if ($key === false) {
        throw new DonaWebhookVerificationError('secret is not base64 after whsec_');
    }
    $expected = base64_encode(hash_hmac('sha256', "{$id}.{$ts}.{$rawBody}", $key, true));
    foreach (explode(' ', $signatures) as $entry) {
        [$version, $sig] = array_pad(explode(',', $entry, 2), 2, '');
        if ($version === 'v1' && $sig !== '' && hash_equals($expected, $sig)) {
            return;
        }
    }
    throw new DonaWebhookVerificationError('no matching signature');
}
